components/tabs/src/storage.rs

```diff
@@ -362,6 +378,32 @@ impl TabsStorage {
     }
 }
 
+// Trim the amount of tabs in a list to fit the specified memory size
+fn trim_tabs_length(
+    tabs: &mut Vec<RemoteTab>,
+    payload_size_max_bytes: usize,
+) -> Result<Vec<RemoteTab>> {
+    // Ported from https://searchfox.org/mozilla-central/rev/84fb1c4511312a0b9187f647d90059e3a6dd27f8/services/sync/modules/util.sys.mjs#422
+    // See bug 535326 comment 8 for an explanation of the estimation
+    let max_serialized_size = (payload_size_max_bytes / 4) * 3 - 1500;
+    let size = compute_serialized_size(tabs)?;
+    if size > max_serialized_size {
+        // Estimate a little more than the direct fraction to maximize packing
+        let cutoff = (tabs.len() * max_serialized_size) / size;
+        tabs.truncate(cutoff);
+
+        // Keep dropping off the last entry until the data fits.
+        while compute_serialized_size(tabs)? > max_serialized_size {
+            tabs.pop();
```

Comment:
nit: This while loop may get expensive if a user has **A LOT** of tabs, since we clear them one at at time, then re-serialize them. but I imagine that's what desktop does too so I see no real practical harm... and any logic that tries to be too smart might get it wrong and over trim

Reply:
I nearly wrote something similar before - where we could do something even funkier by assuming the JSON payload is just an array, and serialize it somewhat manually - eg, serialize each element, and after doing it, see if we can fit it in, and finally put the array payload together manually with something like `format!("[{}]", serialized_tabs.join(",")` or something - then we only serialize each payload once and the ones that don't fit (bar 1) don't get serialized at all! But that's also a bigger change that I think we can optimize later if we care.